src/sensors/can/MG8016EI6.cpp

```diff
@@ -0,0 +1,548 @@
+    write_motor_on();
+}
+
+int MG8016EI6::read(CAN_message_t& msg) {
+    // command IDs that return the base state (temp, torque, speed, angle)
+    // A1, A2, A3, A4, A5, A6, A7, A8, 9C (state 2)
+    // CMD_TORQUE_CONTROL, CMD_SPEED_CONTROL, CMD_MULTI_ANGLE_CONTROL, CMD_MULTI_ANGLE_CONTROL_SPEED_LIMITED,
+    // CMD_ANGLE_CONTROL, CMD_ANGLE_CONTROL_SPEED_LIMITED, CMD_ANGLE_INCREMENT_CONTROL, 
+    // CMD_ANGLE_INCREMENT_CONTROL_SPEED_LIMITED, CMD_READ_STATE_2
+
+    // command IDs that return special state (current, error, etc)
+    // 30, 33, 90, 19, 92 (multi angle), 94 (single angle), 9A/9B (state 1), 9D (state 3)
+    // CMD_READ_PID, CMD_READ_ACCELERATION, CMD_READ_ENCODER, CMD_READ_MULTI_ANGLE,
+    // CMD_READ_ANGLE, CMD_READ_STATE_1, CMD_READ_STATE_3
+
+    // command IDs that dont return anything other than the command echo'd back
+    // 80, 88, 81, 31, 32, 34, 91
+    // CMD_MOTOR_OFF, CMD_MOTOR_ON, CMD_MOTOR_STOP, CMD_WRITE_PID, CMD_WRITE_PID_ROM, CMD_WRITE_ACCELERATION, CMD_WRITE_ENCODER_ZERO
+
+    // early return if msg ID does not match this motor
+    if (msg.id != m_base_id + m_id) {
+        return 0;
+    }
+
+    uint8_t cmd_byte = msg.buf[0];
+
+    // handle the command
+    switch (cmd_byte) {
+    // commands that return the base state
+    case CMD_TORQUE_CONTROL:
+    case CMD_SPEED_CONTROL:
+    case CMD_MULTI_ANGLE_CONTROL:
+    case CMD_MULTI_ANGLE_CONTROL_SPEED_LIMITED:
+    case CMD_ANGLE_CONTROL:
+    case CMD_ANGLE_CONTROL_SPEED_LIMITED:
+    case CMD_ANGLE_INCREMENT_CONTROL:
+    case CMD_ANGLE_INCREMENT_CONTROL_SPEED_LIMITED:
+    case CMD_READ_STATE_2: {
+        // read the state
+        m_state.temperature = msg.buf[1];
+        // torque is given in the signed range [-2048, 2048] which corresponds to [-33A, 33A]
+        // normalize this to [-1, 1]
+        int16_t raw_torque = (int16_t)((msg.buf[3] << 8) | msg.buf[2]);
+        m_state.torque = (float)raw_torque / m_max_torque;
+        
+        // speed is given in 1dps/LSB
+        // need to conver this to rad/s
+        int16_t speed_dps = (int16_t)((msg.buf[5] << 8) | msg.buf[4]);
+        float speed_rad = speed_dps * (PI / 180);
+        m_state.speed = speed_rad;
+        m_state.position = (uint16_t)((msg.buf[7] << 8) | msg.buf[6]);
+        break;
+    }
+    // commands that return special state
+    case CMD_READ_PID: {
+        m_angle_p = msg.buf[2];
+        m_angle_i = msg.buf[3];
+        m_speed_p = msg.buf[4];
+        m_speed_i = msg.buf[5];
+        m_torque_p = msg.buf[6];
+        m_torque_i = msg.buf[7];
+        break;
+    }
+    case CMD_READ_ACCELERATION: {
+        m_acceleration = (int32_t)((msg.buf[7] << 24) | (msg.buf[6] << 16) | (msg.buf[5] << 8) | msg.buf[4]);
+        break;
+    }
+    case CMD_READ_ENCODER: {
+        m_encoder = (uint16_t)((msg.buf[3] << 8) | msg.buf[2]);
+        m_encoder_raw = (uint16_t)((msg.buf[5] << 8) | msg.buf[4]);
+        m_encoder_offset = (uint16_t)((msg.buf[7] << 8) | msg.buf[6]);
```

Comment:
It's a little more obvious to me how all of these buffer operations work- I would still request that you make sure the message buffer structure is described somewhere.